add managing secrets with sops (#3564)
weaveworks/weave-gitops: website/docs/secrets/managing-secrets-with-sops.mdx

---
title: Managing secrets with Sops
hide_title: true
---

# Managing secrets with Sops

Weave GitOps Enterprise now supports managing secrets using [SOPS](https://github.com/mozilla/sops), a tool that encrypts and decrypts secrets using various key management services.

> In this document, we will cover the prerequisites for using SOPS with Weave GitOps Enterprise, and how to configure SOPS for your kubernetes cluster to work with GPG and age keys.

Reviewer: ```suggestion
In this document, we will cover the prerequisites for using SOPS with Weave GitOps Enterprise, and how to configure SOPS for your existing kubernetes cluster to work with GPG and age keys.
```

%no-protection
Key-Type: 1
Key-Length: 4096
Subkey-Type: 1
Subkey-Length: 4096
Expire-Date: 0
Name-Comment: ${KEY_COMMENT}
Name-Real: ${KEY_NAME}
EOF
```

2- Export the key pairs fingerprint in the shell
```bash
gpg --list-secret-keys "${KEY_NAME}"

sec   rsa4096 2020-09-06 [SC]
      710DC0DB6C1662F707095FC30233CB21E656A3CB
      
export KEY_FP="710DC0DB6C1662F707095FC30233CB21E656A3CB"
```

3- Export the generated private key to a kubernetes secret `sops-gpg-private-key` which will be used by flux's kustomize-controller to decrypt the secrets using sops.

```bash
gpg --export-secret-keys --armor "${KEY_FP}" | kubectl create secret generic sops-gpg-private-key --namespace=flux-system --from-file=sops.asc=/dev/stdin
```

4- Export the generated public key to a kubernetes secret `sops-gpg-public-key` which will be used by Weave GitOps Enterprise to encrypt the secrets created from the UI.

```bash
gpg --export --armor "${KEY_FP}" | kubectl create secret generic sops-gpg-public-key --namespace=flux-system --from-file=sops.asc=/dev/stdin
```

> It's recommended to remove the secret from your machine

```bash
gpg --delete-secret-keys "${KEY_FP}"
```

> 5- Create a kustomization for reconciling the secrets on the cluster and set the `--decryption-secret` flag to the name of the private key created in step 3

Reviewer: ```suggestion
5- Create a kustomization for reconciling the secrets on the cluster and set the `--decryption-secret` flag to the name of the private key created in step 3.
```

```bash

> flux create kustomization secrets \

Reviewer: ```suggestion
flux create kustomization gpg-secrets \
```

> --source=secrets \

Reviewer: can you add a comment that this is the gitsoure it will reconcile the secrets from, can be an inline comment

--path=./clusters/my-cluster \
--prune=true \
--interval=10m \
--decryption-provider=sops \
--decryption-secret=sops-gpg-private-key
```

> 6- Annotate the kustomization object created in the previous step with the name and namespace of the public key created in step 4

Reviewer: ```suggestion
6- Annotate the kustomization object created in the previous step with the name and namespace of the public key created in step 4.
```

```bash
kubectl annotate kustomization secrets sops-public-key/name=sops-gpg-public-key sops-public-key/namespace=flux-system
```

The kustomization object should looks like this

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:

>   name: secrets

Reviewer: ```suggestion
  name: gpg-secrets
```

  namespace: flux-system
  annotations:
    sops-public-key/name: sops-gpg-public-key
    sops-public-key/namespace: flux-system
spec:
  interval: 10m
  sourceRef:
    kind: GitRepository
    name: secrets

>   path: ./secrets

Reviewer: ```suggestion
  path: ./secrets/gpg
```

Here are the steps to generate an age key and configure your cluster to work with Weave GitOps Enterprise secrets management.

1- Generate an age key with age-keygen

```bash
age-keygen -o age.agekey

Public key: <public key>
```

2- Export the generated private key to a kubernetes secret `sops-age-private-key` which will be used by flux's kustomize-controller to decrypt the secrets using sops.

```bash
cat age.agekey | kubectl create secret generic sops-age-private-key --namespace=flux-system --from-file=age.agekey=/dev/stdin
```

4- Export the generated public key to a kubernetes secret `sops-age-public-key` which will be used by Weave GitOps Enterprise to encrypt the secrets created from the UI.

```bash
echo "<public key>" | kubectl create secret generic sops-age-public-key --namespace=flux-system --from-file=age.agekey=/dev/stdin
```

> It's recommended to remove the secret from your machine

```bash
rm -f age.ageKey
```

5- Create a kustomization for reconciling the secrets on the cluster and set the `--decryption-secret` flag to the name of the private key created in step 2

```bash

> flux create kustomization secrets \

Reviewer: ```suggestion
flux create kustomization age-secrets \
```

--source=secrets \
--path=./clusters/my-cluster \
--prune=true \
--interval=10m \
--decryption-provider=sops \
--decryption-secret=sops-age-private-key
```

> 6- Annotate the kustomization object created in the previous step with the name and namespace of the public key created in step 4

Reviewer: ```suggestion
6- Annotate the kustomization object created in the previous step with the name and namespace of the public key created in step 4.
```

```bash
kubectl annotate kustomization secrets sops-public-key/name=sops-age-public-key sops-public-key/namespace=flux-system
```

The kustomization object should looks like this

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:

>   name: secrets

Reviewer: ```suggestion
  name: age-secrets
```

  namespace: flux-system
  annotations:
    sops-public-key/name: sops-age-public-key
    sops-public-key/namespace: flux-system
spec:
  interval: 10m
  sourceRef:
    kind: GitRepository
    name: secrets

>   path: ./secrets

Reviewer: ```suggestion
  path: ./secrets/age
```

---
title: Managing secrets with Sops
hide_title: true
---

# Managing secrets with Sops

> Weave GitOps Enterprise now supports managing secrets using [SOPS](https://github.com/mozilla/sops), a tool that encrypts and decrypts secrets using various key management services.

Reviewer 2: Let's remove this one altogether, It feels like an announcement, while it's just documentation. 
```suggestion

```

> In this document, we will cover the prerequisites for using SOPS with Weave GitOps Enterprise, and how to configure SOPS for your existing kubernetes cluster to work with GPG and age keys.

Reviewer 2: ```suggestion
In this document, we will cover the prerequisites for using [SOPS](https://github.com/mozilla/sops) with Weave GitOps Enterprise, and how to configure SOPS for your existing Kubernetes cluster to work with OpenGPG and age keys. For a more advanced setup for SOPS with flux, please refer to this [guide](https://fluxcd.io/flux/guides/mozilla-sops/). 
```

> ## Encrypting secrets using OpenPGP

Reviewer 2: This is OpenGPG, right? 
```suggestion
## Encrypting secrets using OpenGPG
```

> OpenPGP is a way of using Sops to encrypt and decrypt secrets with Weave GitOps Enterprise.

Reviewer 2: ```suggestion
OpenGPG is a way of using Sops to encrypt and decrypt secrets with Weave GitOps Enterprise.
```
same.

> Here are the steps to generate an OpenPGP key and configure your cluster to work with Weave GitOps Enterprise secrets management.

Reviewer 2: ```suggestion
Here are the steps to generate an OpenGPG key and configure your cluster to work with Weave GitOps Enterprise secrets management.
```

```bash
gpg --delete-secret-keys "${KEY_FP}"
```

5- Create a kustomization for reconciling the secrets on the cluster and set the `--decryption-secret` flag to the name of the private key created in step 3.

```bash
flux create kustomization gpg-secrets \
--source=secrets \
--path=./secrets/gpg \
--prune=true \
--interval=10m \
--decryption-provider=sops \
--decryption-secret=sops-gpg-private-key
```

6- Annotate the kustomization object created in the previous step with the name and namespace of the public key created in step 4.

```bash
kubectl annotate kustomization gpg-secrets sops-public-key/name=sops-gpg-public-key sops-public-key/namespace=flux-system
```

The kustomization object should looks like this

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: gpg-secrets
  namespace: flux-system
  annotations:
    sops-public-key/name: sops-gpg-public-key
    sops-public-key/namespace: flux-system
spec:
  interval: 10m
  sourceRef:
    kind: GitRepository
    name: secrets
  path: ./secrets/gpg
  decryption:
    provider: sops
    secretRef:
      name: sops-gpg-private-key
  prune: true
  validation: server

> ```

Reviewer 2: Let's put an obvious note, that state the following: 
```
This is an essential step in order to allow other operators and developers to utilize WeaveGitOps UI to encrypt SOPS secrets using the public key secret in the cluster.
```

```bash
rm -f age.ageKey
```

5- Create a kustomization for reconciling the secrets on the cluster and set the `--decryption-secret` flag to the name of the private key created in step 2

```bash
flux create kustomization age-secrets \
--source=secrets \ # the git source to reconcile the secrets from
--path=./secrets/age \
--prune=true \
--interval=10m \
--decryption-provider=sops \
--decryption-secret=sops-age-private-key
```

6- Annotate the kustomization object created in the previous step with the name and namespace of the public key created in step 4.

```bash
kubectl annotate kustomization secrets sops-public-key/name=sops-age-public-key sops-public-key/namespace=flux-system
```

The kustomization object should looks like this

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: age-secrets
  namespace: flux-system
  annotations:
    sops-public-key/name: sops-age-public-key
    sops-public-key/namespace: flux-system
spec:
  interval: 10m
  sourceRef:
    kind: GitRepository
    name: secrets
  path: ./secrets/age
  decryption:
    provider: sops
    secretRef:
      name: sops-age-private-key
  prune: true
  validation: server
```

> 

Reviewer 2: same here: 

Let's put an obvious note, that state the following:
```
This is an essential step in order to allow other operators and developers to utilize WeaveGitOps UI to encrypt SOPS secrets using the public key secret in the cluster.
```

---
title: Managing secrets with Sops
hide_title: true
---

# Managing secrets with Sops

In this document, we will cover the prerequisites for using [SOPS](https://github.com/mozilla/sops) with Weave GitOps Enterprise, and how to configure SOPS for your existing Kubernetes cluster to work with GPG and age keys.

For a more advanced setup for SOPS with flux, please refer to this [guide](https://fluxcd.io/flux/guides/mozilla-sops/). 

> ## Encrypting secrets using GPG/OpenPGP

Reviewer 3: Did you consider adding in this doc just the bits relevant for Wego, and assuming that key pair exists (for example) and reference those guides (sops and flux documentation)?

Author: I already had this discussion with Sarah and Mostafa 
we found that it's easier for the user to read all the steps in one place specially flux documentation mentions the encryption steps in the same section of the key generation and also it recommends to delete the generated secrets which are needed in this document to export the public key as a kubernetes secret

```bash
flux create kustomization gpg-secrets \
--source=secrets \
--path=./secrets/gpg \
--prune=true \
--interval=10m \
--decryption-provider=sops \
--decryption-secret=sops-gpg-private-key
```

6- Annotate the kustomization object created in the previous step with the name and namespace of the public key created in step 4.

```bash
kubectl annotate kustomization gpg-secrets sops-public-key/name=sops-gpg-public-key sops-public-key/namespace=flux-system
```

The kustomization object should looks like this

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: gpg-secrets
  namespace: flux-system
  annotations:
    sops-public-key/name: sops-gpg-public-key
    sops-public-key/namespace: flux-system
spec:
  interval: 10m
  sourceRef:
    kind: GitRepository
    name: secrets
  path: ./secrets/gpg
  decryption:
    provider: sops
    secretRef:
      name: sops-gpg-private-key
  prune: true
  validation: server
```

> This is an essential step in order to allow other operators and developers to utilize WeaveGitOps UI to encrypt SOPS secrets using the public key secret in the cluster.

## Encrypting secrets using age

> [age](https://github.com/FiloSottile/age) is a simple, modern and secure file encryption tool, It can be used to encrypt secrets using Weave GitOps Enterprise.

Reviewer: This should either be `age is a simple, modern and secure file encryption tool. It can be used to encrypt secrets using Weave GitOps Enterprise.` or
`age is a simple, modern and secure file encryption tool, that can be used to encrypt secrets using Weave GitOps Enterprise.`
But the current punctuation is not correct.

## Security Recommendations

Access to sops decryption secrets should be restricted and allowed only to be read by flux's kustomize controller. This can be done using Kubernetes RBAC.

Here's an example of how you can use RBAC to restrict access to sops decryption secrets:

1. Create a new Kubernetes role that grants read access to sops decryption secrets

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: sops-secrets-role
rules:
- apiGroups: [""]
  resources: ["secrets"]
  resourceNames: ["sops-gpg-private-key", "sops-age-private-key"]
  verbs: ["get", "watch", "list"]
```

2. Bind the role to the service account of the flux's kustomize-controller

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: sops-secrets-rolebinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: sops-secrets-role
subjects:
- kind: ServiceAccount
  name: kustomize-controller
```

> > You would need to ensure that no other rolebindings or clusterrolebndings would allow read the the decryption secret at any time. This could be achieved by leveraging policy capabilities to detect existing and prevent future creation of roles that would grant read secrets permissions. 

Reviewer: ```suggestion
> You would need to ensure that no other rolebindings or clusterrolebndings would allow reading the the decryption secret at any time. This could be achieved by leveraging policy capabilities to detect existing and prevent future creation of roles that would grant read secrets permissions. 
```
